### PR TITLE
Add gauss-newton nonlinear least squares optimization 

### PR DESCRIPTION
**Pull request by amir-saadat**
This is an implementation of gauss-newton algorithm, the simple version of nonlinear optimization algorithms.
I have had a couple of iterations with mblondel and applied some comments on applying jvp and vjp to determine the hessian and gradient wrt params. Also using solve_cg to find the solution to system of linear equations instead of direct inv.

### CODE REVIEW

File: jaxopt/_src/gauss_newton.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+
+class GaussNewtonState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  residual: Any
+  delta: Any
+  error: float
+  gradient: Any
+  jtj: Optional[Any] = None
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:40]: Why do you store `residual`, `delta`, `gradient` and `jtj`? Are they used in underlying applications?

File: jaxopt/_src/gauss_newton.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                             aux=aux)
+
+    return base.OptStep(params=params, state=state)
+
+  def run(self,
+          init_params: Optional[Any] = None,
+          *args,
+          **kwargs) -> base.OptStep:
+    # We override run in order to set init_params=None by default.
+    return super().run(init_params, *args, **kwargs)
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:140]: I would be in favor of making `init_params` mandatory. This way, we don't even need to override `run`.

File: jaxopt/_src/gauss_newton.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Returns:
+      state
+    """
+
+    return GaussNewtonState(iter_num=jnp.asarray(0),
+                            error=jnp.asarray(jnp.inf),
+                            residual=jnp.asarray(jnp.inf),
+                            delta=jnp.asarray(jnp.inf),
+                            gradient=jnp.asarray(jnp.inf),
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:97]: Note that attributes `residual`, `delta`, `gradient` and `jtj` are scalars when returned by `init_state` but are arrays or pytrees when returned by `update`. This will trigger a jit recompilation because the state will change signature. So if you really want to keep these attributes, it may be better to compute their actual values at `init_params` and return that instead.

File: jaxopt/_src/gauss_newton.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+  """
+  residual_fun: Callable
+  maxiter: int = 30
+  tol: float = 1e-5
+  check_bracket: bool = True
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:72]: Not used

File: jaxopt/_src/gauss_newton.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+from typing import Optional
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import implicit_diff as idf
+from jaxopt._src import loop
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:29]: Not used

File: jaxopt/_src/gauss_newton.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    else:
+      params = jnp.asarray(params, float)
+
+    residual, aux = self._fun_with_aux(params, *args, **kwargs)
+
+    matvec = lambda v: self._jtj(params, v, *args, **kwargs)
+    gradient = self._gradient(params, residual, *args, **kwargs)
+
+    delta_params = linear_solve.solve_cg(matvec, gradient)
+    params -= delta_params
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:124]: You can use `tree_sub` here.

File: jaxopt/_src/gauss_newton.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    state = GaussNewtonState(iter_num=state.iter_num + 1,
+                             error=jnp.linalg.norm(delta_params),
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:127]: And `tree_l2_norm` here

File: jaxopt/_src/gauss_newton.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  def run(self,
+          init_params: Optional[Any] = None,
+          *args,
+          **kwargs) -> base.OptStep:
+    # We override run in order to set init_params=None by default.
+    return super().run(init_params, *args, **kwargs)
+
+  def __post_init__(self):
+    # We need this definition in the base solver run function
+    self.optimality_fun = self.residual_fun
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:144]: `optimality_fun` should contain the gradient of `0.5 ||residual_fun(params)||^2`, i.e. `optimality_fun(params) = J.T residual_fun(params)`. So if I'm not wrong it should be:
```python
def optimality_fun(params, *args, **kwargs):
  residual = self.residual_fun(params, *args, **kwargs)
  return self._gradient(self, params, residual, *args, **kwargs)
```

File: jaxopt/_src/gauss_newton.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    else:
+      self._fun = self.residual_fun
+      self._fun_with_aux = lambda *a, **kw: (self.residual_fun(*a, **kw),
+                                             None)
+
+  def _fun_with_args(self, *args, **kwargs):
+    fun_with_args = lambda p: self._fun(p, *args, **kwargs)
+    return self._fun_with_args
+
+  def _jtj(self, params, vec, *args, **kwargs):
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:158]: Maybe add the docstring `"""Product with J.T J"""`?

**Review by mblondel — COMMENTED** [review]
Thanks for your contribution @amir-saadat! I left some comments and questions.

File: jaxopt/_src/gauss_newton.py (hunk 6 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:124]: Good catch, thanks for reminding.

File: jaxopt/_src/gauss_newton.py (hunk 7 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:127]: Fixed. Thanks.

File: jaxopt/_src/gauss_newton.py (hunk 8 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:144]: Yes, that's a great point. This is essentially the RHS of the main equation we are solving to advance the params in each step.

File: jaxopt/_src/gauss_newton.py (hunk 3 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:97]: I implemented your suggestion, very good point, thanks.

File: jaxopt/_src/gauss_newton.py (hunk 2 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:140]: Applied.

File: jaxopt/_src/gauss_newton.py (hunk 1 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:39]: Yes, exactly, they are useful for interpreting the results for the underlying application. for example, if needed, eigenvalues of JT.J can be evaluated. For LM specifically, grad and delta can be useful to determine what cause the algorithm to stop/converge.

**Comment by amir-saadat** [comment]
Thanks @mblondel for the first round of review. They were very helpful and I think I have implemented them now.

File: jaxopt/_src/gauss_newton.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    Given the residual function f(x): R^m -> R^n, `gauss_newton` finds a
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args, **kwargs)
+    ```
+
+  Attributes:
+    residual_fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:57]:
```suggestion
    residual_fun: a smooth function of the form ``residual_fun(x, *args, **kwargs)``.
```

File: jaxopt/_src/gauss_newton.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    maxiter: maximum number of iterations.
+    tol: tolerance.
+    check_bracket: whether to check correctness of the bracketing interval.
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:60]: Remove

File: jaxopt/_src/gauss_newton.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+  def init_state(self,
+                 init_params: Any,
+                 *args,
+                 **kwargs) -> GaussNewtonState:
+    """Initialize the solver state.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``fun``.
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:85]:
```suggestion
      *args: additional positional arguments to be passed to ``residual_fun``.
```

Comment (mblondel) [reply jaxopt/_src/gauss_newton.py:85]: And a few other occurrences.

File: jaxopt/_src/gauss_newton.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Returns:
+      state
+    """
+
+    # Compute actual values of state variables at init_param
+    residual, aux = self._fun_with_aux(init_params, *args, **kwargs)
+    matvec = lambda v: self._jtj_op(init_params, v, *args, **kwargs)
+    gradient = self._jt_op(init_params, residual, *args, **kwargs)
+    # Only materialize J^T.J if verbose is True for debugging
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:95]: I would rather remove `jtj` from the state if possible. I don't think it's a good idea that the state depends on the `verbose` option. If you really need it, I think I would rather add an option `store_jtj` in the constructor.

**Review by mblondel — COMMENTED** [review]
Thanks a lot @amir-saadat. I left a few other comments but overall this looks good! 

It would be great to add an implicit diff test, see e.g. [here](https://github.com/google/jaxopt/blob/main/tests/gradient_descent_test.py#L54) for inspiration. For instance in your `_city_temperature_residual_model` function, you could try to differentiate w.r.t. `x` or `y`.

We will eventually also need documentation and potentially an example but those can be tackled in a separate PR. I think it's better to merge this one first.

File: jaxopt/_src/gauss_newton.py (hunk 13 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:95]: I removed `jtj`. As you mentioned, this becomes more like a debugging feature that I have to think how to extract for the optimizer. Eventually, the eignevalues of the matrix might be of interest. I'll get it into a separate PR once i have an idea.

**Comment by amir-saadat** [comment]
Thanks @mblondel for the nice suggestions. I added the unit test `test_enzyme_reaction_implicit_diff`. It sounded non-trivial conceptually, so please double check that I have done the expected process. Also please let me know how `implicit_diff` is supposed to change the behavior. In the current framework, it doesn't seem to change the results.

File: jaxopt/_src/gauss_newton.py [hunk 14]
```diff
@@ -69,6 +68,7 @@ class GaussNewton(base.IterativeSolver):
   maxiter: int = 30
   tol: float = 1e-5
   verbose: bool = False
+  implicit_diff: bool = False
```
Comment (mblondel) [inline jaxopt/_src/gauss_newton.py:71]: We tend to use `True` as the default in other solvers.

**Review by mblondel — APPROVED** [review]
The new implicit diff test looks good to me. Thanks a lot!

Could you squash your commits? Then, we'll merge!

File: jaxopt/_src/gauss_newton.py (hunk 14 above)
Comment (amir-saadat) [reply jaxopt/_src/gauss_newton.py:71]: Thanks, applied.
